### PR TITLE
Hardening fixes

### PR DESCRIPTION
**Pull request by adamoutler**
# Pull Request: Security Hardening and Proactive Issue Resolution

## Summary
This PR implements critical security hardening measures and proactive user experience enhancements for NetAlertX, focusing on container security, SQL injection prevention, robust testing frameworks, and automated responses to common deployment and configuration issues. As a network monitoring tool handling sensitive device data, these changes significantly reduce the attack surface, improve resilience against exploitation vectors, and provide users with fail-safe mechanisms that prevent broken experiences through early detection and corrective guidance.

### 🔒 Container Security Hardening
**Risk Addressed**: Docker deployments with insecure configurations can expose host networks, lose persistent data, or allow privilege escalation. Common user issues include data loss from improper volume mounting, network isolation failures, and silent configuration errors leading to broken functionality.

**Changes**:
- Identified and implemented critical security checks that halt container operation and display corrective messages in Docker logs, preventing users from experiencing partially broken functionality 
- Added comprehensive Docker integration tests in `test/test_docker_integration.py` to validate secure configurations by systematically testing container configurations against identified and potential failure modes
- Validates `--network=host` requirement for ARP scanning (prevents network isolation bypasses that cause scanning failures)
- Enhanced main README with source build instructions for better user guidance on installation
- Added built-in tests to provide immediate user feedback on secure configuration requirements
- Enforces persistent volume mounts for `/app/config` and `/app/db` (prevents data loss attacks and common user data loss issues)
- Detects tmpfs mounts that could lead to data persistence failures (addresses frequent support requests)
- Automatic configuration seeding on missing files (reduces misconfiguration risks and startup failures)

**Impact**: Prevents unauthorized network access, ensures data integrity in containerized deployments, and eliminates common Docker-related support tickets by failing fast with clear guidance.

### 🛡️ SQL Injection Prevention
**Risk Addressed**: The reporting system was vulnerable to SQL injection attacks through user-controlled filter conditions, potentially allowing unauthorized data access or manipulation.

**Changes**:
- Fixed tests for actual execution within our framework to ensure real-world applicability and catch edge cases
- Converted test framework from unittest to pytest for better security test coverage and maintainability
- Implemented parameterized queries for all dynamic SQL generation to prevent injection attacks
- Added Unicode and edge-case handling to prevent bypass attempts in diverse user environments

**Impact**: Eliminates SQL injection vectors in device filtering and reporting features, protecting against data breaches.

### 🧪 Testing Framework Modernization
**Risk Addressed**: Legacy testing frameworks may miss security edge cases and provide inadequate coverage, leading to undetected vulnerabilities in production.

**Changes**:
- Migrated all test suites to pytest framework for improved reliability and security testing
- Added fixture-based testing for consistent security validation across different scenarios
- Enhanced test coverage for compound condition parsing to catch complex injection attempts
- Implemented automated security regression testing to prevent future vulnerabilities

**Impact**: Ensures ongoing security validation and reduces likelihood of regression vulnerabilities that could affect users.

## Files Changed
- `test/test_docker_integration.py` - New Docker security validation tests addressing common deployment issues
- `test/test_compound_conditions.py` - Converted to pytest, enhanced SQL safety tests
- `test/test_safe_builder_unit.py` - Converted to pytest, expanded injection prevention tests

## Validation Steps
1. **Container Security**: Run Docker integration tests to verify network and storage isolation
2. **SQL Safety**: Execute injection attempt tests to confirm query parameterization
3. **Regression Testing**: All existing functionality preserved with enhanced security

## Risk Assessment
- **Likelihood of Exploitation**: High (network monitoring tools are attractive targets)
- **Impact if Exploited**: Critical (device data exposure, network reconnaissance capabilities)
- **Residual Risk**: Low (defense-in-depth approach with multiple security layers)

## Proactive Issue Resolution
These changes implement a "fail-fast with guidance" approach to common user issues:
- **Early Detection**: Critical security checks halt operations immediately upon detecting misconfigurations, preventing silent failures
- **Clear Messaging**: Corrective messages displayed in Docker logs provide actionable guidance instead of cryptic error messages
- **Automated Validation**: Built-in tests systematically test configurations against known failure modes before deployment
- **User Guidance**: Enhanced documentation and immediate feedback reduce trial-and-error setup attempts
- **Data Protection**: Multiple safeguards prevent data loss scenarios that frequently require support intervention

## Compliance Considerations
These changes align with security best practices for:
- Container security (CIS Docker benchmarks)
- Input validation (OWASP injection prevention)
- Secure coding standards

## Breaking Changes
None. All changes are backward-compatible and enhance existing functionality.

## Future Enhancements
- Adjust database unit tests to run from cold startup (value injection, then read) for more realistic testing scenarios
- Add more edge cases based on user-reported issues to proactively address potential vulnerabilities
- Leverage `/install/production-filesystem/service/scripts/check-*.sh` scripts to halt users on misconfigurations and provide corrective guidance, reducing support overhead and improving user success rates 

## Summary by CodeRabbit

* **New Features**
  * Improved devcontainer/workspace UX: docker socket mount, extra tooling, named persistent volumes, configurable network mode, build timestamp, non-interactive sleep entrypoints, and VS Code task to build the test image.
  * New runtime checks and storage/persistence warnings to surface configuration and permission issues at startup.

* **Bug Fixes**
  * Fixed Python path/env typo and tightened permission/ownership handling with more reliable directory creation.

* **Documentation**
  * README reorganized with expanded TOC and a Quick Start deployment walkthrough.

* **Tests**
  * Added extensive container-focused integration tests and new pytest markers.

### CODE REVIEW

File: install/production-filesystem/services/scripts/check-first-run-config.sh [hunk 1]
```diff
@@ -5,23 +5,22 @@
...
-    >&2 printf "%s" "${CYAN}"
-    >&2 cat <<EOF
+    >&2 cat <<'EOF'
 ══════════════════════════════════════════════════════════════════════════════
 🆕  First run detected. Default configuration written to ${NETALERTX_CONFIG}/app.conf.
 
     Review your settings in the UI or edit the file directly before trusting
     this instance in production.
 ══════════════════════════════════════════════════════════════════════════════
 EOF
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply install/production-filesystem/services/scripts/check-first-run-config.sh]: Suggestion Applied. Thanks.

File: install/production-filesystem/services/scripts/check-root.sh [hunk 2]
```diff
@@ -20,11 +20,16 @@ if [ "${CURRENT_UID}" -eq 0 ]; then
       * Keep the default USER in the image (20211:20211), or
       * In docker-compose.yml, remove any 'user:' override that sets UID 0.
 
+	Note: As a courtesy, this special mode is only used to set the permissions
+	of /app/db and /app/config to be owned by the netalertx user so future 
+	runs work correctly.
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply install/production-filesystem/services/scripts/check-root.sh]: Tabs versus spaces will pwn some messages.

File: install/production-filesystem/services/start-backend.sh [hunk 3]
```diff
@@ -11,5 +11,5 @@ done
 # Force kill if graceful shutdown failed
 killall -KILL python3 &>/dev/null
 
-echo "python3 $(cat /services/config/python/backend-extra-launch-parameters 2>/dev/null) -m server > >(tee /app/log/stdout.log) 2> >(tee /app/log/stderr.log >&2)"
-exec python3 $(cat /services/config/python/backend-extra-launch-parameters 2>/dev/null) -m server > >(tee /app/log/stdout.log) 2> >(tee /app/log/stderr.log >&2)
+echo "python3 $(cat /services/config/python/backend-extra-launch-parameters 2>/dev/null) -m server > /app/log/stdout.log 2> >(tee /app/log/stderr.log >&2)"
+exec python3 $(cat /services/config/python/backend-extra-launch-parameters 2>/dev/null) -m server > /app/log/stdout.log 2> >(tee /app/log/stderr.log >&2)
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply install/production-filesystem/services/start-backend.sh]: This makes it more difficult to read overall.  Instead of being a command line that someone can read and follow, it would be a set of variables which increases cognative load during review and troubleshooting.  This is an entrypoint and shoudl be simple.

File: server/api_server/nettools_endpoint.py [hunk 4]
```diff
@@ -80,7 +81,7 @@ def speedtest():
     try:
         # Run speedtest-cli command
         result = subprocess.run(
-            [f"{INSTALL_PATH}/back/speedtest-cli", "--secure", "--simple"],
+            ["speedtest-cli", "--secure", "--simple"],
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply server/api_server/nettools_endpoint.py]: @coderabbitai , this is using whatever the use has on their system.  speedtest-cli is only ever a command line interface for SpeedTest.net. The entire container in a secure environment is read-only, but the unsecure environment is not. In the secure environment this path would be /opt/venv/bin/speedtest-cli, however, it may also be /bin/, /usr/bin/, /opt/myenv or others when a user installs locally on a system outside of docker.   What do you recommend for adapting to all the various scenarios?

File: test/docker_tests/test_container_environment.py [hunk 5]
```diff
@@ -0,0 +1,951 @@
...
+        mounts_ls += f" {target}"
+    mounts_ls += " || true; echo '--- END MOUNTS ---'; \n"
+
+    script = (
+        mounts_ls
+        + f"sh /entrypoint.sh & pid=$!; "
+        + f"sleep {sleep_seconds}; "
+        + "if kill -0 $pid >/dev/null 2>&1; then kill -TERM $pid >/dev/null 2>&1 || true; fi; "
+        + "wait $pid; code=$?; if [ $code -eq 143 ]; then exit 0; fi; exit $code"
+    )
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply test/docker_tests/test_container_environment.py:246]: Good catch. It's not an FString so we dont' need to worry about it.

File: test/test_compound_conditions.py [hunk 6]
```diff
@@ -5,322 +5,326 @@
...
-import unittest
+import pytest
 from unittest.mock import MagicMock
 
 # Mock the logger module before importing SafeConditionBuilder
 sys.modules['logger'] = MagicMock()
 
 # Add parent directory to path for imports
-sys.path.insert(0, '/tmp/netalertx_hotfix/server/db')
+sys.path.insert(0, '/workspaces/NetAlertX')
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply test/test_compound_conditions.py]: Will try this. Performing unit testing.

[the remark this answers is not in the document]
Comment (adamoutler) [reply test/test_compound_conditions.py]: This file was only modified because it caused other tests to fail. I take no responsibility for this file. It ruins other unit testing.

**Comment by adamoutler** [comment]
Coderabbit is a bit overly verbose suddenly. It's good to have more info but it takes effort to scroll to the bottom of each comment.

File: install/production-filesystem/services/scripts/check-capabilities.sh [hunk 7]
```diff
@@ -27,5 +27,6 @@ then
 ══════════════════════════════════════════════════════════════════════════════
 EOF
     >&2 printf "%s" "${RESET}"
-    exit 1
-fi
\ No newline at end of file
+    exit 0
+fi
+exit 0
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply install/production-filesystem/services/scripts/check-capabilities.sh]: Added note.

File: install/production-filesystem/services/scripts/check-app-permissions.sh [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+if [ "$(id -u)" -eq 0 ]; then
+    echo "Running as root. Ensuring correct ownership and permissions..."
+
+    # Set ownership to netalertx user and group for all read-write paths
+    chown -R netalertx:netalertx ${READ_WRITE_PATHS}
+
+    # Set directory and file permissions for all read-write paths
+    find ${READ_WRITE_PATHS} -type d -exec chmod 700 {} +
+    find ${READ_WRITE_PATHS} -type f -exec chmod 600 {} +
+fi
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply install/production-filesystem/services/scripts/check-app-permissions.sh:47]: Applied fix.

File: install/production-filesystem/services/scripts/check-first-run-config.sh [hunk 9]
```diff
@@ -5,23 +5,22 @@
 if [ ! -f ${NETALERTX_CONFIG}/app.conf ]; then
     mkdir -p "${NETALERTX_CONFIG}" || {
         >&2 echo "ERROR: Failed to create config directory ${NETALERTX_CONFIG}"
-        exit 1
+        exit 0
     }
     cp /app/back/app.conf "${NETALERTX_CONFIG}/app.conf" || {
         >&2 echo "ERROR: Failed to copy default config to ${NETALERTX_CONFIG}/app.conf"
-        exit 1
+        exit 0
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply install/production-filesystem/services/scripts/check-first-run-config.sh]: If there's no config and it can't be copied, the user cannot start the system and will wonder what's wrong.  exiting non-zero will terminate the startup process. Change applied.

File: install/production-filesystem/services/scripts/check-nonpersistent-storage.sh [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    ${label} relies on host storage to persist data across container restarts.
+    Mount this directory from the host or a named volume before trusting the
+    container's output.
+
+    Example:
+        --mount type=bind,src=/path/on/host,dst=${path}
+══════════════════════════════════════════════════════════════════════════════
+EOF
+    >&2 printf "%s" "${RESET}"
+}
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply install/production-filesystem/services/scripts/check-nonpersistent-storage.sh:34]: accepted. Thanks.

File: test/docker_tests/test_container_environment.py [hunk 11]
```diff
@@ -0,0 +1,951 @@
...
+    in read-only containers. Tests each required mount point when missing.
+    Expected: "Write permission denied" error with path, guidance to add volume mounts.
+    """
+    paths = _setup_mount_tree(tmp_path, "missing_mount_app_config")
+    volumes = _build_volume_args(paths, skip={"app_config"})
+    result = _run_container("missing-mount-app-config", volumes, user="20211:20211")
+    _assert_contains(result.stdout, "Write permission denied")
+    _assert_contains(result.stdout, "/app/api")
+    assert result.returncode != 0
+
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply test/docker_tests/test_container_environment.py:735]: Looks like we were checking the wrong paths in the output.  Should be fixed now.

File: README.md [hunk 12]
```diff
@@ -38,6 +46,14 @@ docker run -d --rm --network=host \
...
 ```
 
+To deploy a containerized instance directly from the source repository, execute the following BASH sequence:
+```bash
+git clone https://github.com/jokob-sk/NetAlertX.git
+cd NetAlertX
+docker compose up --force-recreate --build
+# To customize: edit docker-compose.yaml and run that last command again
+```
+
```
Comment (jokob-sk) [inline README.md:56]: Let's add a note that this is without persistent storage - e.g. all your config and your db will be wiped on container rebuild if you don't mount persistent storage.

Comment (adamoutler) [reply README.md:56]: It's stored on the host, in Docker Named Volumes

```shell
root@HLAB-A25:/var/lib/docker/volumes/netalertx_netalertx_config/_data# ls -al
total 792
drwx------ 2 20211 20211   4096 Oct 26 11:28 .
drwx-----x 3 root  root    4096 Oct 25 22:13 ..
-rw-r--r-- 1 20211 20211 159665 Oct 26 11:28 app.conf
-rw-r--r-- 1 20211 20211   3963 Oct 26 11:24 app.conf_2025-10-26_16-24-16.backup
-rw-r--r-- 1 20211 20211 159709 Oct 26 11:25 app.conf_2025-10-26_16-25-08.backup
-rw-r--r-- 1 20211 20211 159709 Oct 26 11:25 app.conf.bak
-rw------- 1 20211 20211 159704 Oct 26 11:24 app.conf_old_setting_names_20251026112435.bak
-rw-r--r-- 1 20211 20211 159660 Oct 26 11:28 app.conf_old_setting_names_20251026112826.bak
root@HLAB-A25:/var/lib/docker/volumes/netalertx_netalertx_config/_data# ls  ../../netalertx_netalertx_db/_data/
app.db  app.db-shm  app.db-wal
root@HLAB-A25:/var/lib/docker/volumes/netalertx_netalertx_config/_data# docker volume ls
DRIVER    VOLUME NAME
local     netalertx_netalertx_config
local     netalertx_netalertx_db
local     vscode
```

File: .devcontainer/scripts/setup.sh [hunk 13]
```diff
@@ -102,6 +143,12 @@ configure_source() {
         killall python3 &>/dev/null
         sleep 0.2
     done
+	sudo chmod 777 /opt/venv/lib/python3.12/site-packages/ && \
+    sudo chmod 005 /opt/venv/lib/python3.12/site-packages/
+	sudo chmod 666 /var/run/docker.sock
+
+	echo "      -> Updating build timestamp"
+	date +%s > ${NETALERTX_FRONT}/buildtimestamp.txt
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply .devcontainer/scripts/setup.sh:151]: Fixed

File: test/docker_tests/dockerfiles/no-error-compose.yml [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+    logging:
+      driver: "json-file"
+      options:
+        max-size: "10m"
+        max-file: "3"
+    restart: unless-stopped
+
+volumes:
+  netalertx_config_test:
+  netalertx_db_test:
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply test/docker_tests/dockerfiles/no-error-compose.yml]: Removed test file

File: install/production-filesystem/services/scripts/check-ramdisk.sh [hunk 15]
```diff
@@ -42,9 +42,9 @@ warn_if_not_dedicated_mount "${NETALERTX_API}"
 warn_if_not_dedicated_mount "${NETALERTX_LOG}"
 
 if [ "${failures}" -ne 0 ]; then
-    exit 1
+    exit 0
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply install/production-filesystem/services/scripts/check-ramdisk.sh]: Not all checks need to error, which prevents access to the app.   This is one of those.

File: .devcontainer/scripts/setup.sh [hunk 16]
```diff
@@ -102,6 +143,12 @@ configure_source() {
         killall python3 &>/dev/null
         sleep 0.2
     done
+	sudo chmod 777 /opt/venv/lib/python3.12/site-packages/ && \
+  sudo chmod 005 /opt/venv/lib/python3.12/site-packages/
+	sudo chmod 666 /var/run/docker.sock
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply .devcontainer/scripts/setup.sh:148]: This is a decision which keeps things easier to work with in the single-user devcontainer environment.  This is not production code and this reduces complexity when dealing with different user IDs and different environments.

File: install/production-filesystem/services/scripts/update_vendors.sh [hunk 17]
```diff
@@ -19,7 +19,7 @@ TEMP_FILE="/services/run/tmp/ieee-oui.txt.tmp"
 OUTPUT_FILE="/services/run/tmp/ieee-oui.txt"
 
 # Download the file using wget to stdout and process it
-if ! wget --timeout=30 --tries=3 "https://standards-oui.ieee.org/oui/oui.txt" -O /dev/stdout | \
+if ! wget --timeout=30 --tries=3 "https://standards-oui.ieee.org/oui/oui.txt" -O /dev/stdout 2>/dev/null | \
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply install/production-filesystem/services/scripts/update_vendors.sh:22]: We don't need this message.  If the environment is offline then this just won't update.

File: install/production-filesystem/services/start-php-fpm.sh [hunk 18]
```diff
@@ -24,8 +22,8 @@ done
 trap cleanup EXIT
 trap forward_signal INT TERM
 
-echo "/usr/sbin/php-fpm83 -y \"${PHP_FPM_CONFIG_FILE}\" -F >>\"${LOG_APP_PHP_ERRORS}\" 2>&1 &"
-/usr/sbin/php-fpm83 -y "${PHP_FPM_CONFIG_FILE}" -F >>"${LOG_APP_PHP_ERRORS}" 2>&1 &
+echo "Starting /usr/sbin/php-fpm83 -y \"${PHP_FPM_CONFIG_FILE}\" -F >>\"${LOG_APP_PHP_ERRORS}\" 2>/dev/stderr &"
+/usr/sbin/php-fpm83 -y "${PHP_FPM_CONFIG_FILE}" -F >>"${LOG_APP_PHP_ERRORS}" 2> /dev/stderr &
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply install/production-filesystem/services/start-php-fpm.sh:26]: Intentional. Php logs are handled in the config

File: install/production-filesystem/services/scripts/check-persistent-storage.sh [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+        --mount type=bind,src=/path/on/host,dst=${path}
+
+        # Named volume
+        --mount type=volume,src=netalertx-data,dst=${path}
+
+    Apply one of these mount options and restart the container.
+══════════════════════════════════════════════════════════════════════════════
+EOF
+    >&2 printf "%s" "${RESET}"
+}
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply install/production-filesystem/services/scripts/check-persistent-storage.sh:67]: No.  Not all scripts need to terminate the container

File: install/production-filesystem/services/scripts/check-nonpersistent-storage.sh [hunk 20]
```diff
@@ -0,0 +1,45 @@
+#!/bin/sh
+# check-storage-extra.sh - ensure additional NetAlertX directories are persistent mounts.
+
+
+if [ "${NETALERTX_DEBUG}" == "1" ]; then
+    exit 0
+fi
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply install/production-filesystem/services/scripts/check-nonpersistent-storage.sh:7]: It's set in Dockerfile.

File: .vscode/tasks.json [hunk 21]
```diff
@@ -160,5 +160,26 @@
...
+      "problemMatcher": [],
+      "group": {
+        "kind": "build",
+        "isDefault": false
+      },
+      "icon": {
+        "id": "beaker",
+        "color": "terminal.ansiBlue"
+      }
+    }
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply .vscode/tasks.json:183]: Fixed

File: install/production-filesystem/entrypoint.sh [hunk 22]
```diff
@@ -49,25 +49,47 @@ printf '
...
+        echo "${script_name}: FAILED with ${FAILED_STATUS}"
+        echo "Failure detected in: ${script}"
+    fi
 done
 
+
+if [ ${FAILED_STATUS} ]; then
+    echo "Container startup checks failed with exit code ${FAILED_STATUS}."
+    exit ${FAILED_STATUS}
+fi
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply install/production-filesystem/entrypoint.sh:81]: Applied.

File: install/production-filesystem/services/scripts/check-app-permissions.sh [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply install/production-filesystem/services/scripts/check-app-permissions.sh:47]: It's unquoted because it's multiple paths.  Quoting it would cause a problem.

File: test/docker_tests/test_container_environment.py [hunk 24]
```diff
@@ -0,0 +1,926 @@
...
+    3. Missing Required Mounts: Simulates forgetting to mount required persistent volumes
+    in read-only containers. Tests each required mount point when missing.
+    Expected: "Write permission denied" error with path, guidance to add volume mounts.
+    """
+    paths = _setup_mount_tree(tmp_path, "missing_mount_app_log")
+    volumes = _build_volume_args(paths, skip={"app_log"})
+    result = _run_container("missing-mount-app-log", volumes, user="20211:20211")
+    _assert_contains(result, "Write permission denied", result.args)
+    _assert_contains(result, "/app/api", result.args)
+
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply test/docker_tests/test_container_environment.py:749]: Applied fixes. thanks.

File: server/api_server/nettools_endpoint.py [hunk 25]
```diff
@@ -2,12 +2,37 @@
...
+    if not os.access(path, os.X_OK):
+        raise RuntimeError(f"speedtest-cli found at {path} but is not executable")
+    return path
+
+try:
+    SPEEDTEST_CLI_PATH = _get_speedtest_cli_path()
+except Exception as e:
+    # Warn but don't crash import — the endpoint will return 503 when called.
+    print(f"Warning: {e}", file=sys.stderr)
+    SPEEDTEST_CLI_PATH = None
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply server/api_server/nettools_endpoint.py:34]: No.  This is something that still needs review by primary maintainer, and this isn't a plugin.

File: test/docker_tests/test_container_environment.py [hunk 26]
```diff
@@ -0,0 +1,926 @@
...
+        )
+        config_file.chmod(0o600)
+    if seed_db:
+        db_file = paths["app_db"] / "app.db"
+        shutil.copyfile(
+            "/workspaces/NetAlertX/db/app.db",
+            db_file,
+        )
+        db_file.chmod(0o600)
+
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply test/docker_tests/test_container_environment.py:146]: Sure.  It doesn't seem to break anything.

File: test/docker_tests/test_container_environment.py [hunk 27]
```diff
@@ -0,0 +1,950 @@
...
+    stdouterr = (
+        re.sub(r'\x1b\[[0-9;]*m', '', result.stdout or '') +
+        re.sub(r'\x1b\[[0-9;]*m', '', result.stderr or '')
+    )   
+    result.output = stdouterr
+    # Print container output for debugging in every test run.
+    try:
+        print("\n--- CONTAINER out ---\n", result.output)
+    except Exception:
+        pass
```
[the remark this answers is not in the document]
Comment (adamoutler) [reply test/docker_tests/test_container_environment.py:269]: @coderabitai, 
There is no need for this

File: docker-compose.yml [hunk 28]
```diff
@@ -63,27 +66,26 @@ services:
...
-      LISTEN_ADDR: 0.0.0.0                          # Listen for connections on all interfaces
-      PORT: ${PORT}                                 # Application port
-      ALWAYS_FRESH_INSTALL: ${ALWAYS_FRESH_INSTALL} # Set to true to reset your config and database on each container start
-      NETALERTX_DEBUG: 0                            # 0=kill all services and restart if any dies. 1 keeps running dead services.
-      TZ: ${TZ}                                     # Timezone, e.g. Europe/Paris
-      # APP_CONF_OVERRIDE={"SCAN_SUBNETS":"['192.168.1.0/24 --interface=eth1']","GRAPHQL_PORT":"20223","UI_theme":"Light"} # (optional) app.conf settings override
-      # LOADED_PLUGINS=["DHCPLSS","PIHOLE","ASUSWRT","FREEBOX"] # (optional) default plugins to load
+      LISTEN_ADDR: ${LISTEN_ADDR:-0.0.0.0}                      # Listen for connections on all interfaces
+      PORT: ${PORT:-20211}                               # Application port
+      GRAPHQL_PORT: ${GRAPHQL_PORT:-20212}                       # GraphQL API port
```
Comment (jokob-sk) [inline docker-compose.yml]: this needs to be changed to APP_CONF_OVERRIDE={"GRAPHQL_PORT":"20212"}

Comment (adamoutler) [inline docker-compose.yml]:
```suggestion
      APP_CONF_OVERRIDE: ${GRAPHQL_PORT:-20212}                       # GraphQL API port
```

Comment (adamoutler) [reply docker-compose.yml]: Ok.

**Review by adamoutler — COMMENTED** [review]
Done.

File: docker-compose.yml [hunk 29]
```diff
@@ -63,27 +66,26 @@ services:
...
-      LISTEN_ADDR: 0.0.0.0                          # Listen for connections on all interfaces
-      PORT: ${PORT}                                 # Application port
-      ALWAYS_FRESH_INSTALL: ${ALWAYS_FRESH_INSTALL} # Set to true to reset your config and database on each container start
-      NETALERTX_DEBUG: 0                            # 0=kill all services and restart if any dies. 1 keeps running dead services.
-      TZ: ${TZ}                                     # Timezone, e.g. Europe/Paris
-      # APP_CONF_OVERRIDE={"SCAN_SUBNETS":"['192.168.1.0/24 --interface=eth1']","GRAPHQL_PORT":"20223","UI_theme":"Light"} # (optional) app.conf settings override
-      # LOADED_PLUGINS=["DHCPLSS","PIHOLE","ASUSWRT","FREEBOX"] # (optional) default plugins to load
+      LISTEN_ADDR: ${LISTEN_ADDR:-0.0.0.0}                      # Listen for connections on all interfaces
+      PORT: ${PORT:-20211}                               # Application port
+      APP_CONF_OVERRIDE: ${GRAPHQL_PORT:-20212}                       # GraphQL API port
```
Comment (jokob-sk) [inline docker-compose.yml:71]: Thanks! Unfortunately still not correct :) APP_CONF_OVERRIDE is a JSON object,so I assume it should be: 

APP_CONF_OVERRIDE=${APP_CONF_OVERRIDE:-{"GRAPHQL_PORT":"20212"}}

not sure about the `-`

Comment (adamoutler) [reply docker-compose.yml:71]: `APP_CONF_OVERRIDE: ${GRAPHQL_PORT:-20212}`  sets  NetAlertX env variable APP_CONF_OVERRIDE to 20212 unless GRAPHQL_PORT is set. 

* no variable: `APP_CONF_OVERRIDE=20212`
* Variable: `APP_CONF_OVERRIDE=${GRAPHQL_PORT}`

The `:-` annotation sets the fall back.  Tell me how it should be.

Comment (jokob-sk) [reply docker-compose.yml:71]: Then I think this is the right syntax:

`APP_CONF_OVERRIDE=${APP_CONF_OVERRIDE:-{"GRAPHQL_PORT":"20212"}}`

note APP_CONF_OVERRIDE is a JSON object, e.g. `{"GRAPHQL_PORT":"20212"}` or `{"GRAPHQL_PORT":"20212","UI_LANG":"English (en_us)"}`

no variable should result into:

`APP_CONF_OVERRIDE={"GRAPHQL_PORT":"20212"}`

hope that makes sense

Comment (adamoutler) [reply docker-compose.yml:71]: I think I'd rather set that inside the container, in the `entrypoint sh` and test it.  Will test tomorrow.

Comment (adamoutler) [reply docker-compose.yml:71]: Ok, so if you set GRAPHQL_PORT: 20212, then it will use that.  If you set APP_CONF_OVERRIDE it will not use GRAPHQL_PORT and ignore it.  APP_CONF_OVERRIDE will override GRAPHQL_PORT.

File: install/production-filesystem/entrypoint.sh [hunk 30]
```diff
@@ -49,25 +49,58 @@ printf '
...
 set -u
 
-# Run all pre-startup checks to validate container environment and dependencies
+FAILED_STATUS=""
+echo "Startup pre-checks"
 for script in ${SYSTEM_SERVICES_SCRIPTS}/check-*.sh; do
-	sh "$script"
+    if [ -n "${DISABLE_STARTUP_CHECKS:-}" ]; then
+        echo "Skipping startup checks as DISABLE_STARTUP_CHECKS is set."
+        break
```
[the remark this answers is not in the document]
Comment (jokob-sk) [reply install/production-filesystem/entrypoint.sh]: Is this a valid finding?
